ballista/scheduler/src/state/task_manager.rs

```diff
@@ -523,57 +656,57 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> TaskManager<T, U>
         }
     }
 
-    #[allow(dead_code)]
     pub async fn prepare_task_definition(
         &self,
         task: TaskDescription,
     ) -> Result<TaskDefinition> {
         debug!("Preparing task definition for {:?}", task);
 
-        let job_id = task.partition.job_id.clone();
-        let stage_id = task.partition.stage_id;
+        let job_id = task.partitions.job_id.clone();
+        let stage_id = task.partitions.stage_id;
 
-        if let Some(mut job_info) = self.active_job_cache.get_mut(&job_id) {
-            let plan = if let Some(plan) = job_info.encoded_stage_plans.get(&stage_id) {
-                plan.clone()
-            } else {
-                let mut plan_buf: Vec<u8> = vec![];
-                let plan_proto = U::try_from_physical_plan(
-                    task.plan,
-                    self.codec.physical_extension_codec(),
-                )?;
-                plan_proto.try_encode(&mut plan_buf)?;
-
-                job_info
-                    .encoded_stage_plans
-                    .insert(stage_id, plan_buf.clone());
-
-                plan_buf
-            };
+        let props = self
+            .state
+            .get_session(&task.session_id)
+            .await?
+            .state()
+            .config_options()
+            .entries()
+            .into_iter()
+            .filter_map(|ConfigEntry { key, value, .. }| {
+                value.map(|value| KeyValuePair { key, value })
+            })
+            .collect();
+
+        let jobs = self.active_job_queue.jobs();
+
+        if let Some(_job_info) = jobs.get_mut(&job_id) {
+            let optimizer = OptimizeTaskGroup::new(task.partitions.partitions.clone());
```

Comment:
I wonder if we do pruning instead, we wouldn't need `CoalesceTasksExec` but can use `CoalescePartitionsExec`.

Reply:
Yeah, `CoaalesceTasksExec` is basically copied from `CoalescePartitionsExec`. I think it could make sense to just update `CoalescePartitionsExec` in DF to take a `Vec<usize>` of partitions so we can use it directly. 

But I thought about doing it that way and ultimately decided against it because pruning is a bit hard to generalize. To deal with user-defined `ExecutionPlan` implementations such as `ArchiveExec` you need to either change the `AsPhysicalPlan` trait or introduce a new trait that handles pruning.